### PR TITLE
Refactor database queries, take #2

### PR DESCRIPTION
**Pull request by thomaseizinger**
Fix up SQL queries

This PR does a few things:
* cleans up the SQL to make the queries clearer in terms of intent, as well as eliminating the use of an extra transaction in some write queries.
* adds some additional testing
* (mostly) eliminates the use of `serde_json::to_string()`, making the data columns behave in a more sane manner

### CODE REVIEW

File: daemon/src/db.rs [hunk 1]
```diff
@@ -1,61 +1,63 @@
...
+        order
+            .creation_timestamp
+            .duration_since(SystemTime::UNIX_EPOCH)?
+            .as_secs() as i64,
+    )
+    .bind(
+        order
+            .creation_timestamp
+            .duration_since(SystemTime::UNIX_EPOCH)?
+            .subsec_nanos() as i32,
```
Comment (da-kami) [inline daemon/src/db.rs:55]: ❓ Just out of curiosity: Are nanos actually represented as `i32` or why are we not using `i64` here like for `secs`?

File: daemon/src/db.rs [hunk 2]
```diff
@@ -66,97 +68,83 @@ pub async fn load_order_by_id(
...
-    let cfd_state = serde_json::to_string(&cfd.state)?;
-
-    // save cfd + state in a transaction to make sure the state is only inserted if the cfd was
-    // inserted
+        insert into cfds (
+            order_id,
+            order_uuid,
+            quantity_usd
+        )
+        select
```
Comment (da-kami) [inline daemon/src/db.rs:122]: ❓ Why are the selects after the insert needed? Is this here to check that the insert worked - i.e. testing it on the fly?

Interesting pattern - I find it adds some complexity when looking at the coee but don't mind much keeping that.

File: daemon/src/db.rs [hunk 3]
```diff
@@ -183,7 +171,7 @@ pub async fn append_cfd_state(cfd: &Cfd, conn: &mut PoolConnection<Sqlite>) -> a
         insert into cfd_states (
             cfd_id,
             state
-        ) values (?, ?);
+        ) values ($1, $2);
```
Comment (da-kami) [inline daemon/src/db.rs:172]: ❓ if we have validation `select`s above, why not here as well?

File: daemon/src/db.rs [hunk 4]
```diff
@@ -246,159 +232,205 @@ pub async fn load_cfd_by_order_id(
...
+        cfd as (
+            select
+                ord.order_id,
+                id as cfd_id,
+                quantity_usd
+            from cfds
+                inner join ord on ord.order_id = id
+        ),
+
+        state as (
```
Comment (da-kami) [inline daemon/src/db.rs:264]: nit: I think this should be called `cfds_with_latest_state` - because in my understanding that is what this represents.

File: daemon/src/db.rs [hunk 5]
```diff
@@ -246,159 +232,205 @@ pub async fn load_cfd_by_order_id(
...
+            select
+                id as state_id,
+                cfd.order_id,
+                cfd.quantity_usd,
+                state
+            from cfd_states
+                inner join cfd on cfd.cfd_id = cfd_states.cfd_id
+            where id in (
```
Comment (da-kami) [inline daemon/src/db.rs:272]: ❓ Why is this `where id in` and not `where id = `? 
`max` should always return exactly one, can sqlite not handle that?

File: daemon/src/db.rs [hunk 6]
```diff
@@ -246,159 +232,205 @@ pub async fn load_cfd_by_order_id(
...
-            orders.creation_timestamp as creation_timestamp,
-            orders.term as term,
-            orders.oracle_event_id,
-            cfds.quantity_usd as quantity_usd,
-            cfd_states.state as state
-        from cfds as cfds
-        inner join orders as orders on cfds.order_id = orders.id
-        inner join cfd_states as cfd_states on cfd_states.cfd_id = cfds.id
-        where cfd_states.state in (
+        with ord as (
```
Comment (da-kami) [inline daemon/src/db.rs:235]: nit: could we name these placeholders like the tables that they represent? 
I find it hard to map from table to placeholder to next statement (...).

File: daemon/src/db.rs [hunk 7]
```diff
@@ -246,159 +232,205 @@ pub async fn load_cfd_by_order_id(
...
+            ord.term_nanos as "term_nanos: i32",
+            ord.origin as "origin: crate::model::cfd::Origin",
+            ord.oracle_event_id,
+            state.quantity_usd,
+            state.state
+
+        from ord
+            inner join state on state.order_id = ord.order_id
+
+        where ord.uuid = $1
```
Comment (da-kami) [inline daemon/src/db.rs:301]: Not part of this PR, but thought I mention it:

I was unable to figure out a nice way to reuse statement code. These huge statements basically do the same thing besides the final `where` clause`. Can you think of a way to re-use the code - do you think it would be a good idea?

**Review by da-kami — APPROVED** [review]
Looking good, just a few questions and minor suggestions.

File: daemon/src/db.rs [hunk 8]
```diff
@@ -408,88 +440,119 @@ pub async fn load_cfds_by_oracle_event_id(
     oracle_event_id: BitMexPriceEventId,
     conn: &mut PoolConnection<Sqlite>,
 ) -> anyhow::Result<Vec<Cfd>> {
-    let oracle_event_id = oracle_event_id.to_string();
-
+    let event_id = oracle_event_id.to_string();
```
Comment (da-kami) [inline daemon/src/db.rs:441]: I extended the tests for `load_all_cfds` - but given that these other statements duplicate the complex join SQL code we might also want to add more tests for loading by `oracle_event_id` and `order_id` to make sure we cover someone only changing one of these statements. (unless we can achieve partly re-using the SQL statement code).

File: daemon/src/db.rs (hunk 7 above)
Comment (thomaseizinger) [reply daemon/src/db.rs:301]: The only thing I can think of is adding SQL views to the DB.

File: daemon/src/db.rs (hunk 1 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:55]: `i32`, yes. And not for completely arbitrary reasons either!

```rust
/// A span of time with nanosecond precision.
///
/// Each `Duration` is composed of a whole number of seconds and a fractional part represented in
/// nanoseconds.
///
/// This implementation allows for negative durations, unlike [`core::time::Duration`].
#[derive(Clone, Copy, Default, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Duration {
    /// Number of whole seconds.
    seconds: i64,
    /// Number of nanoseconds within the second. The sign always matches the `seconds` field.
    nanoseconds: i32, // always -10^9 < nanoseconds < 10^9
    #[allow(clippy::missing_docs_in_private_items)]
    padding: Padding,
```

File: daemon/src/db.rs (hunk 2 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:122]: It is in fact [very standard SQL](https://www.w3schools.com/sql/sql_insert_into_select.asp). Basically, you're pointing the output data of the `SELECT` query to be the input data for the `INSERT` query.

File: daemon/src/db.rs (hunk 7 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:301]: FWIW I've tried this already with no luck. It may be possible, but it's not simple nor obvious.

File: daemon/src/db.rs (hunk 3 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:172]: Because we're not returning anything back to `rust` here.

File: daemon/src/db.rs (hunk 6 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:235]: In this specific case, no. This is on account of the fact that `ORDER` is a protected keyword in SQL.

File: daemon/src/db.rs (hunk 5 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:272]: Because the `GROUP BY` clause does the grouping, and the `MAX()` operation is applied to the group. Equality might work, didn't try. `MAX()` is explicitly what we want on a per-group basis.

File: daemon/src/db.rs (hunk 3 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:172]: Er, scratch that, misread the question. The reason is because the function previously had `#[allow(dead_code)]`, so it was simply ignored. You're correct, should fix that.

File: daemon/src/db.rs (hunk 2 above)
Comment (da-kami) [reply daemon/src/db.rs:122]: cool, I did not know that!

File: daemon/src/db.rs (hunk 7 above)
Comment (da-kami) [reply daemon/src/db.rs:301]: I was pretty sure you had tried already :) - guess it's not worth to invest more time. It might be a good idea to extend the tests of `load_cfds_by_oracle_event_id` and `load_cfds_by_order_id` to be sure we don't into trouble if we ever change the statements. Either way, approved.

File: daemon/src/db.rs (hunk 6 above)
Comment (thomaseizinger) [reply daemon/src/db.rs:235]: I like this idea.

File: daemon/src/db.rs [hunk 9]
```diff
@@ -408,88 +438,119 @@ pub async fn load_cfds_by_oracle_event_id(
...
     Ok(cfds)
 }
 
+fn convert_to_system_time(row_secs: i64, row_nanos: i32) -> Result<SystemTime> {
+    let secs = row_secs as i128;
+    let nanos = row_nanos as i128;
+    let offset_dt = OffsetDateTime::from_unix_timestamp_nanos(secs * 1_000_000_000 + nanos)?;
+
+    Ok(SystemTime::from(offset_dt))
+}
```
Comment (thomaseizinger) [inline daemon/src/db.rs:552]: Can you create an issue to get rid of the nanoseconds for all these timestamps? Not worth the complexity IMO.

File: daemon/src/model.rs [hunk 10]
```diff
@@ -109,7 +110,17 @@ pub struct WalletInfo {
...
+    Clone,
+    Copy,
+    SerializeDisplay,
+    DeserializeFromStr,
+    PartialEq,
+    Eq,
+    Hash,
+    PartialOrd,
+    Ord,
+    sqlx::Type,
```
Comment (thomaseizinger) [inline daemon/src/model.rs]: Adding this without `transparent` or another annotation does not do anything on structs as far as I know.

File: daemon/src/db.rs (hunk 8 above)
Comment (thomaseizinger) [reply daemon/src/db.rs:441]: Agreed!

**Review by thomaseizinger — COMMENTED** [review]
Thank you @DeliciousHair! I can't request changes on my own PR, so I'll just post a comment.

Couple of things:

- The commit message needs some more work to follow https://chris.beams.io/posts/git-commit/.
- Daniel made a few good comments, please address them.
- There is an unnecessary derive on `BitMexPriceEventId`
- I would like a follow-up issue on the nanosecond topic.

File: daemon/src/db.rs (hunk 9 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:552]: The reason it was left in is I am unclear as to how the timestamps are used initially, so that one that is created via `Systemtime::now()` and one retrieved from the DB will be unable to be matched without them, **if** that's a thing. In other words, the extra complexity here would be (potentially) replaced with the opposite complexity somewhere else.

Can do this of course, just need to be sure that it won't create issues somewhere else.

Comment (thomaseizinger) [reply daemon/src/db.rs:552]: Yep that makes a lot of sense. It needs to be thought through, hence it would be good to have a follow-up issue!

File: daemon/src/db.rs (hunk 8 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:441]: Added to `test_multiple_cfd_updates_per_cfd()`

File: daemon/src/db.rs (hunk 4 above)
Comment (da-kami) [reply daemon/src/db.rs:264]: was this not actioned on purpose? would be nice to say so. Resolved for me means that it was actioned.

**Review by da-kami — APPROVED** [review]

**Review by thomaseizinger — COMMENTED** [review]
Looks good!

Did you do a manual test as well?

**Comment by DeliciousHair** [comment]
> Did you do a manual test as well?

Yes, in addition to more EDA-style ad-hoc testing prior to the switch to an in-memory DB. All seems well-fine to me now, fingers crossed!

File: daemon/src/db.rs (hunk 4 above)
Comment (DeliciousHair) [reply daemon/src/db.rs:264]: While there is no general style guide that I am aware of for SQL in general (could be well wrong about this, I'm no expert!) there certainly isn't one for us internally. That being said, in practice most SQL aliases end up with really useful aliases like `a`, `b`, etc., which obviously suck but in my experience shorter phrases > longer ones as they tend to compound and things get progressively less readable with larger, more complex queries (which this most certainly is not, of course).

Thus, my action was to ignore the suggestion as I simply disagree with it is all; I think `state` is good alias as it captures the intent of the sub-query, that being that it is getting the "state" part of the overall query.

Apologies if you disagree, I read the above as a suggestion only is all, and since it didn't affect anything beyond stylistic concerns I chose to keep my original wording.

Comment (da-kami) [reply daemon/src/db.rs:264]: all good, would just be nice to add a quick comment next time and then resolve the comment - then I know that it was on purpose and not just accidentally resolved.
